routing/probe.go

```diff
@@ -0,0 +1,514 @@
+package routing
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/plprobelab/go-kademlia/kad"
+	"github.com/plprobelab/go-kademlia/kaderr"
+	"github.com/plprobelab/go-kademlia/key"
+	"github.com/plprobelab/go-kademlia/util"
+)
+
+type RoutingTableCpl[K kad.Key[K], N kad.NodeID[K]] interface {
+	kad.RoutingTable[K, N]
+
+	// Cpl returns the longest common prefix length the supplied key shares with the table's key.
+	Cpl(kk K) int
+
+	// CplSize returns the number of nodes in the table whose longest common prefix with the table's key is of length cpl.
+	CplSize(cpl int) int
+}
+
+// The Probe state machine performs regular connectivity checks for nodes in a routing table.
+//
+// The state machine is notified of a new entry in the routing table via the [EventProbeAdd] event. This adds the node
+// to an internal list and sets a time for a check to be performed, based on the current time plus a configurable
+// interval.
+//
+// Connectivity checks are performed in time order, so older nodes are processed first. The connectivity check performed
+// is the same as for the Include state machine: ask the node for closest nodes to itself and confirm that the node
+// returns at least one node in the list of closer nodes. The state machine emits the [StateProbeConnectivityCheck]
+// state when it wants to check the status of a node.
+//
+// The state machine expects to be notified either with the [EventProbeMessageResponse] or the
+// [EventProbeMessageFailure] events to determine the outcome of the check. If neither are received within a
+// configurable timeout the node is marked as failed.
+//
+// Nodes that receive a successful response have their next check time updated to the current time plus the configured
+// check interval.
+//
+// Nodes that fail a connectivity check, or are timed out, are removed from the routing table and from the list of nodes
+// to check. The state machine emits the [StateProbeNodeFailure] state to notify callers of this event.
+//
+// The state machine accepts a [EventProbePoll] event to check for outstanding work such as initiating a new check or
+// timing out an existing one.
+//
+// The [EventProbeRemove] event may be used to remove a node from the check list and from the routing table.
+//
+// The state machine accepts the [EventProbeNotifyConnectivity] event as a notification that an external system has
+// performed a suitable connectivity check, such as when the node responds to a query. The probe state machine treats
+// these events as if a successful response had been received from a check by advancing the time of the next check.
+type Probe[K kad.Key[K], A kad.Address[A]] struct {
+	rt RoutingTableCpl[K, kad.NodeID[K]]
+
+	// nvl is a list of nodes with scoring values
```

Comment:
"scoring values" are just the check deadlines?

Reply:
The full design adds scores for nodes based on number of queries participated in (or other signals) and then eviction from the routing table would take that into account. For example very low scoring nodes  might be evicted even if they pass a connectivity check, or very high scoring nodes might be retained with a drop in score that would keep them longer in case they start responding again. The details need to be worked out, but this is the state machin where I expect that scoring to be done.

Reply:
Nice! Thanks for clarifying.